Add client testing via containers

ehelms opened the pull request:
Adds client testing by spinning up containers defined in `variables.json` based on `foreman_version`, `clients` and `hostname` of the server. This design currently relies on the idea that there is a dedicated box that runs smoker that can install hammer on it and run at least podman.

Design considerations:
 * Using apypie instead of hammer to be able to more all in one style testing
 * Relies on the tests living all in a single module and is order dependent
 * Handles creating and destroying requested containers to allow for easy re-running of tests and single, pytest driven command to do so

ehelms commented on tests/test_client.py:
```diff
@@ -0,0 +1,186 @@
+
+
+@pytest.fixture
+def entities():
```
Should be centralized at some point as other tests we want to migrate here will use these.

ekohl replied:
We already use https://pypi.org/project/pytest-variables/. That is probably the best way since it also allows overriding things.

ehelms replied:
Given these are not things that should change, I wouldn't want to have to create them via variables every time. Perhaps a future use case will steer it that direction.

ekohl replied:
I thought that we could include a default variables file in the repo and always load that. Doesn't look like that's currently possible, but we can probably override `pytest_configure` in `conftest.py` to always inject `defaults.json` (or whatever name) as a first value.

ehelms commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+
+def pytest_generate_tests(metafunc):
+    with open(metafunc.config.option.variables[0]) as file:
+      variables = json.load(file)
```
The set of clients, and other dynamic input parameters rely on using the variables fixture as input. That way, one can lay down a file with all of the configuration needed and run it for the whole suite.

ekohl replied:
Odd indenting
```suggestion
        variables = json.load(file)
```

ehelms commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+      metafunc.parametrize('client', variables['clients'], indirect=True, scope='session')
+
+def teardown_module():
+    podman_ids = subprocess.check_output(['podman', 'ps', '-q', '--filter', "name=smoker-client"]).decode().strip()
```
Relies on setting the name of the container with the `smoker-client` prefix to allow us to find and remove them all at the end of each run. This helps keep the tests self contained and easily re-runnable.

ehelms commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+
+@pytest.fixture(scope='session')
+def smoker():
+    return testinfra.get_host("local://localhost")
```
Represents the smoker box that commands like hammer are ran on.

ekohl replied:
In my test I could use `local://` as a connection string.

ehelms commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+def pytest_generate_tests(metafunc):
+    with open(metafunc.config.option.variables[0]) as file:
+      variables = json.load(file)
+      metafunc.parametrize('client', variables['clients'], indirect=True, scope='session')
```
It would be nice to additionally run these matrixed over the server_hostname as well. I am not sure entirely how to do that yet (haven't tried). Such that we get a test of each client against the server, and each client against a foreman proxy.

ekohl replied:
No need do this while the file is open
```suggestion
    metafunc.parametrize('client', variables['clients'], indirect=True, scope='session')
```

ehelms commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+
+@pytest.mark.katello_agent
+def test_host_is_registered(client, smoker):
+    hostname = client.check_output('hostname -s')
```
I should turn this into a function

ehelms commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+
+def test_install_katello_host_tools(client, foreman_version):
+    release = client.system_info.release.split('.')[0]
+    client.run_expect([0], "yum -y install http://yum.theforeman.org/client/%s/el%s/x86_64/foreman-client-release.rpm" % (foreman_version, release))
```
Need a better place to do this to help support other distributions and spins that have different client repositories.

ekohl replied:
I'd suggest a `foreman_client_release_rpm_url` fixture.

ehelms commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+    release = client.system_info.release.split('.')[0]
+
+    if release != 6 or release != 7:
+        pytest.skip("katello-agent not available on EL8")
```
There are better ways to do this, and a good candidate for the server to inform through the API what is present.

ekohl replied:
I think a fixture would work:
```python
@pytest.fixture
def katello_agent_client(client):
    release = client.system_info.release.split('.')[0]
    if release not in (6, 7):
        pytest.skip(f"katello-agent not available on EL{release}")
    return client
```

Then in all katello agent tests you would use this fixture.

evgeni replied:
Or with the new `os_release` fixture:

```python
@pytest.fixture
def katello_agent_client(katello_client, os_release):
    if os_release not in ("6", "7"):
        pytest.skip(f"katello-agent not available on EL{release}")
    return katello_client
```

(you can't do pytest.skipIf on fixtures :()

evgeni replied:
And now I wonder if this particular test is even really meaningful or if the fixture should do `yum install k-agent` and then  the other `test_katello_agent_*` tests can rely on that?

ehelms commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+
+
+@pytest.mark.container
+def test_fetch_container_content(client, smoker, entities, server_hostname):
```
This test is kind of ugly. As you will note I am currently testing this on smoker not the client. Running podman inside podman is difficult without using a `--privileged` container.

ekohl replied:
How about [skopeo](https://github.com/containers/skopeo):
```console
# mkdir containers
# skopeo copy docker://registry.fedoraproject.org/fedora:latest dir:containers
Getting image source signatures
Copying blob 8fde7942e775 done
Copying config 79fd58dc76 done
Writing manifest to image destination
Storing signatures
# ls containers/
79fd58dc76113dac76a120f22cadecc3b2d1794b414f90ea368cf66096700053  8fde7942e775327f2d6ddaf08f45fda74201a7769951b0413860e21b59b4bf82  manifest.json  version
```

Also has an option to log in. There are also options to inspect them from a remote source so you don't need to download it all.

ehelms replied:
That suggestion worked.

ekohl commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+@pytest.fixture(scope='session')
+def client(request):
+    name = "smoker-client-%s" % request.param.replace('/', '_').replace(':', '_')
+    podman_id = subprocess.check_output(['podman', 'ps', '-q', '--filter', "name=%s" % name])
```
In https://github.com/theforeman/smoker/pull/10 I used a local connection with the podman module to see if it was running.

ekohl commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+    return variables['foreman_version']
+
+def pytest_generate_tests(metafunc):
+    with open(metafunc.config.option.variables[0]) as file:
```
File is a built in function. Do not shadow it.

ekohl commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+      variables = json.load(file)
+      metafunc.parametrize('client', variables['clients'], indirect=True, scope='session')
+
+def teardown_module():
```
Better than a teardown module is to do it as part of the fixture. https://docs.pytest.org/en/stable/fixture.html#fixture-finalization-executing-teardown-code

ehelms replied:
Yea, that was my original implementation. As soon as I introduced the parameterization, that stopped working. Even when I try to specify `session` or `module` based scope the code after the `yield` in the client fixture gets called after every test which destroys the container between tests. I tried many things, perhaps someone else might have better luck. This worked nicely.

ekohl replied:
For what it's worth: for good test isolation, I think you should delete the container between runs.

I noticed that current Fedora and Debian don't include systemd in their container and it wouldn't surprise me if a future CentOS wouldn't do the same thing. Perhaps we need to prepare some Dockerfiles and build our own base images (locally) which also include `subscription-manager`. That way you have a guaranteed fresh container without any state. You could launch the container with `--rm` to make sure it's gone after it stops.

ehelms replied:
I see it as trying to walk that line between building expensive, fully isolated test cases vs. orchestrating and asserting across a setup to reduce the time to execute. Seems we can:

 a) have small dedicated tests with 1-2 assertions but are intentionally order dependent (similar to bats today)
 b) have fewer tests but multiple assertions within said test to group testing concerns but reduce repeated setup
 c) have fully isolated, dedicated tests with setup and teardown that do not aim to minimize run time or re-use

I believe that a/b are similar to the acceptance tests in the Puppet modules in style.

ekohl replied:
> I believe that a/b are similar to the acceptance tests in the Puppet modules in style.

I think rspec mostly follows the order as well but our contexts aren't really dependent on them.

What mostly matters is what happens when you use python-xdist. Looks like now we're not using it, but if you run them with `pytest -n 4`. pytest is more aggressive in assuming tests are independent.

There is https://smarie.github.io/python-pytest-steps/ but I have no experience with it and it also states it's untested with python-xdist.

I would prefer to use option c with well chosen fixtures to minimize run time. For example, a fixture that gives you a running container where subscription-manager is already present. When using podman, that can cache the image to reduce the time needed.

ekohl commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+    podman_ids = subprocess.check_output(['podman', 'ps', '-q', '--filter', "name=smoker-client"]).decode().strip()
+    podman_ids = podman_ids.split("\n")
+
+    for id in podman_ids:
```
`id` is a built in function. Do not shadow it.

ekohl commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+
+
+@pytest.mark.katello_agent
+def test_katello_agent_package_install(client, smoker):
```
I think this test depends on `katello-agent` being installed. I don't think you can really depend on another test and if we use xdist to run tests concurrently this can easily break. I'd suggest to do the installation as part of the fixture so you know for sure it's installed.

ekohl commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+    podman_id = subprocess.check_output(['podman', 'ps', '-q', '--filter', "name=%s" % name])
+
+    if not podman_id:
+        subprocess.check_output(['podman', 'run', '--name', name, '-dt', request.param]).decode().strip()
```
In my tests I gave them all a shared label so you could easily query them.

ekohl commented on tests/test_client.py:
```diff
@@ -0,0 +1,184 @@
+def test_register_subscription_manager_with_username_and_password(client, entities, server_hostname):
+    client.run("yum erase -y 'katello-ca-consumer-*'")
+
+    client.run_expect([0], "rpm -Uvh http://%s/pub/katello-ca-consumer-latest.noarch.rpm" % server_hostname)
```
I'd introduce a fixture for the consumer rpm url. If we have multiple, we can also use `pytest_generate_tests` to create a matrix.

ekohl reviewed (commented):
The dependencies that exist within the tests do concern me.

ehelms commented:
Latest round of updates:

 * addresses code comments
 * adds support for katello-agent (systemd based service) tests to run
 * adds a custom built centos:6 for testing

Two things I am seeing running the tests right now:

 * Intermittent failures with EL6 registration via activation key
 * katello-agent package removal test failing

ekohl commented on images/centos6/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+RUN rm -f /etc/yum.repos.d/*
+ADD centos-6-vault.repo /etc/yum.repos.d/centos-6-vault.repo
+ADD epel-subscription-manager.repo /etc/yum.repos.d/epel-subscription-manager.repo
+RUN yum -y install https://archives.fedoraproject.org/pub/archive/epel/6/x86_64/epel-release-6-8.noarch.rpm
```
Isn't EPEL6 itself empty now? Does this still make sense to include?

ehelms commented:
Tis not empty, just in the archives. This is where python-qpid packages are
for EL6

On Fri, Dec 4, 2020, 5:38 PM Ewoud Kohl van Wijngaarden <
notifications@github.com> wrote:

> *@ekohl* commented on this pull request.
> ------------------------------
>
> In images/centos6/Dockerfile
> <https://github.com/theforeman/smoker/pull/11#discussion_r536425463>:
>
> > @@ -0,0 +1,8 @@
> +FROM registry.centos.org/centos:6
> +
> +RUN rm -f /etc/yum.repos.d/*
> +ADD centos-6-vault.repo /etc/yum.repos.d/centos-6-vault.repo
> +ADD epel-subscription-manager.repo /etc/yum.repos.d/epel-subscription-manager.repo
> +RUN yum -y install https://archives.fedoraproject.org/pub/archive/epel/6/x86_64/epel-release-6-8.noarch.rpm
>
> Isn't EPEL6 itself empty now? Does this still make sense to include?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/smoker/pull/11#pullrequestreview-545344694>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACHT4ZX34L5E77LLAZJSXDSTFQIFANCNFSM4UM36EMQ>
> .
>

ehelms commented:
Now has dropped the hammer dependency and thus the dependency on the underlying box being setup and configured to talk to the server. This makes use of Apypie to handle communication via the API. Additionally, I switched it to `"{}".format` syntax. I could still use some feedback on design question at https://github.com/theforeman/smoker/pull/11#discussion_r536360202

ekohl commented:
> Additionally, I switched it to `"{}".format` syntax.

I think we can also depend on Python 3.6+ so you can use f-strings. So `f"value: {value}"` works (instead of `"value: {}".format(value)`)

ehelms commented:
Also noticing, if I try to run the entire test suite, the client tests are skipped when `--base-url` is included. Almost as if the `pytest_generate_test` function is interfering across files, though I thought it was isolated to the file

ekohl commented:
I think we need to move that to `conftest.py` and centralize the `pytest_generate_test` method.

ehelms commented:
> I think we need to move that to `conftest.py` and centralize the `pytest_generate_test` method.

We have different sets of tests being parameterized in different ways. Is that possible with centralized pytest_generate_test ? Does this need more an approach of running the test suites through individual `pytest` calls and a warpper around that? (e.g. rake test:unit, rake test:acceptance)

ekohl commented on tests/test_client.py:
```diff
@@ -0,0 +1,230 @@
+        pytest.skip("katello-agent not available on EL8")
+
+    client.run_expect([0], 'yum -y install katello-agent')
+    time.sleep(30)
```
Why is there a sleep here?

ehelms commented:
Another round of updates moving to allow each test to run individually (granted they all require `fb-katello-content.bats` to run, but we can address that later) and move some setup into containers and some into fixtures. I think this has reached a good point to add value to our tests. Introduced `conftest.py` to centralize the parameterization generation and some fixtures.

ekohl commented on tests/conftest.py:
```diff
@@ -0,0 +1,86 @@
+
+    if 'url' in metafunc.fixturenames:
+        base_url = f"https://{variables['server_hostname']}"
+        assert base_url
```
This will always be true since `base_url` is always a non-empty string. It have raised an exception on the line before, but if it gets here, it'll be truthy.

evgeni commented on images/centos7/Dockerfile:
```diff
@@ -0,0 +1,5 @@
+FROM registry.centos.org/centos:7
+
+RUN yum -y install http://yum.theforeman.org/client/$foreman_version/el7/x86_64/foreman-client-release.rpm
```
doesn't `foreman_version` need to be defined as an `ARG`?

ehelms replied:
It does not for this work. I can add it if we prefer.

evgeni replied:
If it works, it works.

ekohl commented on tests/conftest.py:
```diff
@@ -0,0 +1,86 @@
+
+
+def pytest_generate_tests(metafunc):
+    with open(metafunc.config.option.variables[0]) as f:
+        variables = json.load(f)
+    assert variables
```
I think this accesses the raw option. Note that pytest-variables also accepts other file formats, such as YAML. It also supports layering which you're breaking here. That's why I used `metafunc.config._variables` to get the user object. Not great either, but you don't break functionality.

ehelms replied:
Sorry, I do not understand what raw option or breaking layering means here.

evgeni replied:
I think Ewoud means the fact that `pytest-variables` can (and does) load variables itself. And it can load from multiple files, merging the result: https://github.com/pytest-dev/pytest-variables#specifying-variables

So I think you should be able to use

```python
variables = metafunc.config._variables  # pylint: disable=protected-access
```

evgeni commented on requirements.txt:
```diff
@@ -2,3 +2,5 @@ pytest-base-url
 pytest-variables
 pytest-xdist
+pytest-testinfra
+apypie
```
so it *was* a good decision to keep it separate and not fold it into FAM completely ;-)

ehelms replied:
Yea, yea. :+1:

ekohl commented on tests/conftest.py:
```diff
@@ -0,0 +1,86 @@
+        username=variables['username'],
+        password=variables['password'],
+        api_version=2,
+        verify_ssl=True,
```
Here SSL verification is on, but to get the menu I turned it off. From a remote machine there's no guarantee that the certificate is present. Perhaps this should be configurable, but at least match within smoker.

evgeni replied:
I'd expect this needs to be `False`? Or do you add the cert to the global cert store before?

ekohl commented on tests/test_client.py:
```diff
@@ -0,0 +1,185 @@
+    subprocess.check_call(['podman', 'rm', '-f', container_id])
+
+
+def wait_for_task(task, api):
+    duration = 60
+    poll = 15
+
+    while task['state'] not in ['paused', 'stopped']:
+        duration -= poll
+        if duration <= 0:
+            break
+        time.sleep(poll)
+
+        task = api.resource('foreman_tasks').call('show', {'id': task['id']})
```
@evgeni is this something we should provide in apypie? Perhaps it could also be done smarter in the future. For example, Foreman could give hints about the time to back off.

Could also be interesting to think about asyncio here.

evgeni replied:
Well, apypie has no concept of tasks, so in a very clean room, I'd say: no, it doesn't belong there.

on the other hand, this looks copied from FAM's foreman_helper.py, so maybe it's time to provide a `apypie.foreman` submodule?

evgeni commented on tests/test_client.py:
```diff
@@ -0,0 +1,185 @@
+
+    image_id = subprocess.check_output(['podman', 'images', '-q', '--filter', f"reference={request.param}"])
+    if not image_id:
+        subprocess.check_output(['sh', 'build_images.sh', foreman_version])
```
check_call is sufficient here, as we don't care about the output?

ekohl replied:
This means that it can run `build_images.sh` for every client that's not found. Perhaps it should be structured as `build_image.sh` and pass an argument to avoid this?

ekohl commented on tests/conftest.py:
```diff
@@ -0,0 +1,86 @@
+def api(variables):
+    api = apypie.Api(
+        uri=f"https://{variables['server_hostname']}",
+        username=variables['username'],
+        password=variables['password'],
```
Please use the `user` fixture. It has a layer on top of it (admin/changeme) and keeps it consistent.

evgeni commented on tests/test_client.py:
```diff
@@ -0,0 +1,185 @@
+
+@pytest.mark.katello_agent
+def test_install_katello_agent(katello_client):
+    release = katello_client.system_info.release.split('.')[0]
```
should this be a fixture, given you use it in multiple tests?

ekohl commented on tests/conftest.py:
```diff
@@ -0,0 +1,86 @@
+    assert variables
+
+    if 'url' in metafunc.fixturenames:
+        base_url = f"https://{variables['server_hostname']}"
```
This should use the [existing base-url](https://pypi.org/project/pytest-base-url/) that we pass.

ehelms replied:
I'm effectively dropping the use of `--base-url` in favor of configuring all dynamic inputs through `variables.json` rather than splitting them across two methods of specification.

ekohl replied:
And I disagree with that. Mostly because you break the existing tests. pytest-selenium depends on pytest-base-url and uses it.

ekohl commented on variables.json.example:
```diff
@@ -0,0 +1,11 @@
+{
+  "server_hostname": "foreman.example.com",
```
:-1: on this. If needed, it can be parsed from the base_url.

ehelms replied:
See https://github.com/theforeman/smoker/pull/11#discussion_r539310678

ehelms commented on tests/test_client.py:
```diff
@@ -0,0 +1,172 @@
+    container = testinfra.get_host("podman://" + container_id)
+
+    container.run_expect([0], f"rpm -Uvh http://{registration_hostname}/pub/katello-ca-consumer-latest.noarch.rpm")
+    container.run_expect([0], f"subscription-manager register --force --org={entities['organization_label']} --activationkey=\"{entities['activation_key']}\"")
```
While the tests are isolated and run independently now, these two steps here combined take about a minute. That means every test has a runtime of at least a minute. So 10 tests, 3 OSes, that's around half an hour. Previously, when the tests were considered to be order dependent I was getting 6 minute run times. 

On the plus side, I was able to run it with `-n 4` with a run time of 4 minutes.

ehelms commented:
Rebased. I added a provision so that if a set of clients is not defined those tests will be skipped entirely. This should allow us to get this change in, and then make updates to our deployments to configure it and trigger the tests. This will also allow some iteration and tuning of the tests. Given this sets the stage for doing Apypie based tests I would like to get it in as soon as I can to start iterating on replacing more bats tests that are API driven.

evgeni commented on tests/test_client.py:
```diff
@@ -0,0 +1,167 @@
+import subprocess
+
+
+@pytest.fixture(scope='session')
```
should this still be scoped to the session? I think each iteration should get a new client.

ehelms replied:
The problem I can into was that this fixture requires other `session` fixtures and pytest complained. So I have to make this a session or de-session all the rest of the fixtures. I am happy to go either way.

evgeni replied:
I think I'd prefer this not being `session` and thus the others to follow, as `session` implies caching of fixtures, and we really want a new container for every test.

@ekohl what do you think?

evgeni commented on tests/conftest.py:
```diff
@@ -0,0 +1,85 @@
+        metafunc.parametrize('url', pages)
+
+    if 'katello_client' in metafunc.fixturenames:
+        if 'clients' not in variables:
+            clients = []
+        else:
+            clients = variables['clients']
```
```suggestion
        clients = variables.get('clients', [])
```

ekohl commented on build_images.sh:
```diff
@@ -0,0 +1,7 @@
+
+foreman_version=$1
+
+podman build images/centos8 -t smoker-test/centos8:$foreman_version --build-arg foreman_version=$foreman_version
```
Food for thought: should we use labels to indicate capabilities? Like a katello-agent label. That way in the future we can add Debian clients where there is no katello-agent, but we can test content. Not something that needs to be implemented now

ekohl commented on build_images.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
Should we use `-e` to stop execution if something fails?

ekohl commented on build_images.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+foreman_version=$1
```
Just to be sure, please check if it's empty
```suggestion
foreman_version=$1
if [[ -z $foreman_version ]] ; then
  echo "Usage: $0 FOREMAN_VERSION"
  exit 1
fi
```

ekohl commented on requirements.txt:
```diff
@@ -2,3 +2,5 @@ pytest-base-url
 pytest-selenium
 pytest-variables
 pytest-xdist
+pytest-testinfra
```
Please keep the file sorted

ekohl commented on tests/conftest.py:
```diff
@@ -0,0 +1,82 @@
+        uri=base_url,
+        username=user.username,
+        password=user.password,
+        api_version=2,
```
Isn't this the default? Shouldn't we rely on that?

evgeni replied:
No, it's default for Foreman, but not for apipie (it has no way to learn what the default is).

ehelms replied:
What happens if the default changes in the future and then this breaks? I mostly borrowed this from FAM, so @evgeni may have an additional reason.

ekohl commented on tests/test_client.py:
```diff
@@ -0,0 +1,167 @@
+    image_id = subprocess.check_output(['podman', 'images', '-q', '--filter', f"reference={request.param}"])
+    assert image_id
+
+    container_id = subprocess.check_output(['podman', 'run', '-v', '/dev/log:/dev/log', '--systemd=true', '-dt', request.param, '/sbin/init']).decode().strip()
```
Rather than passing `/sbin/init`, shouldn't we set `CMD` in the Dockerfiles we maintain? Also, `--systemd=true` is the default value. I'd also prefer long options in scripts. So something like this:
```suggestion
    container_id = subprocess.check_output(['podman', 'run', '--volume', '/dev/log:/dev/log', '--detach', '--tty', request.param]).decode().strip()
```

ekohl commented on variables.json.example:
```diff
@@ -0,0 +1,10 @@
+{
+  "foreman_version": "nightly",
```
This is no longer used

ehelms commented:
I renamed `test_client.py` to `test_content_client.py` for now as that paves the way to have a dedicated `test_puppet_client.py` as a follow up to this using the same container concepts.

evgeni commented on tests/test_content_client.py:
```diff
@@ -0,0 +1,175 @@
+    assert entities['errata_id'] in errata_ids
+
+
+@pytest.mark.katello_agent
+def test_install_katello_agent(katello_client, os_release):
+    if os_release == "8":
+        pytest.skip("katello-agent not available on EL8")
+
+    katello_client.run_expect([0], 'yum -y install katello-agent')
+    katello_agent = katello_client.package('katello-agent')
+
+    assert katello_agent.is_installed
+
+
+@pytest.mark.katello_agent
+def test_katello_agent_package_install(katello_client, api, os_release):
+    if os_release == "8":
+        pytest.skip("katello-agent not available on EL8")
+
+    katello_client.run_expect([0], 'yum -y install katello-agent')
+    host = foreman_host(katello_client, api)
+    task = api.resource('host_packages').call('install', {'host_id': host['id'], 'packages': ['gorilla']})
+    wait_for_task(task, api)
+
+    gorilla = katello_client.package('gorilla')
+    assert gorilla.is_installed
```
I don't think `test_install_katello_agent` -- it only tests that k-agent can be installed, but we already ensured the client can install packages locally. How about converting this to a fixture, and then re-using it for the other k-agent tests, like this:

```suggestion
@pytest.fixture
def katello_agent_client(katello_client, os_release):
    if os_release not in ("6", "7"):
        pytest.skip(f"katello-agent not available on EL{os_release}")

    katello_client.run_expect([0], 'yum -y install katello-agent')
    
    katello_agent = katello_client.package('katello-agent')
    assert katello_agent.is_installed
    
    return katello_client
    

@pytest.mark.katello_agent
def test_katello_agent_package_install(katello_agent_client, api):
    host = foreman_host(katello_client, api)
    task = api.resource('host_packages').call('install', {'host_id': host['id'], 'packages': ['gorilla']})
    wait_for_task(task, api)

    gorilla = katello_client.package('gorilla')
    assert gorilla.is_installed
```

ehelms commented:
Added a fixture for katello-agent and updated the parsing of the variables in the generate function. Note, this currently does not pass all tests but thats due to bugs in Katello that are being addressed. As this is coded not to run until we define clients it should be safe to merge.

evgeni commented:
this needs a rebase and flake8 complains, but overall I think this is at a state where I'd say "MERGE IT" and see how it progresses from there.

@ekohl  raised a good idea about labels, definitely something we should look at in the future.

Another thing that came into my mind: is there any time benefit of pre-building those containers and uploading them to quay.io instead of building them every time locally? I admit the major part of time is probably spent in fetching the base container, not installing sub-man, but I still wanted to ask.

ehelms commented:
> Another thing that came into my mind: is there any time benefit of pre-building those containers and uploading them to quay.io instead of building them every time locally? I admit the major part of time is probably spent in fetching the base container, not installing sub-man, but I still wanted to ask.

I have considered this as a future enhancement to optimize. The only potential gotcha is different versions of the client repositories as those are set at build time based on an argument to the build.

ekohl commented on tests/conftest.py:
```diff
@@ -74,6 +82,20 @@ def user(variables):
     return _user(variables)
 
 
+@pytest.fixture(scope='session')
+def entities():
+    return {
+        'organization_label': 'Test_Organization',
+        'activation_key': 'Test AK',
+        'product': 'Test Product',
+        'product_label': 'Test_Product',
+        'yum_repository': 'Zoo',
+        'yum_repository_label': 'Zoo',
+        'errata_id': 'RHEA-2012:0055',
+        'container_repository_label': 'foremanbusybox'
+    }
```
I'm debating if these should rather be individual fixtures. I can't quite make up my mind about them. Time will tell whether this is a good choice or if we want to change it at some point. A typical example where we're still learning the best practices.

ehelms commented on tests/test_content_client.py:
```diff
@@ -0,0 +1,163 @@
+
+@pytest.fixture
+def katello_client(request, entities, registration_hostname):
+    image_id = subprocess.check_output(['podman', 'images', '-q', '--filter', f"reference={request.param}"])
+    assert image_id
```
Yes

ekohl commented on tests/test_content_client.py:
```diff
@@ -0,0 +1,169 @@
+    assert image_id
+
+    try:
+        container_id = subprocess.check_output([
+            'podman',
+            'run',
+            '--volume',
+            '/dev/log:/dev/log',
+            '--detach',
+            '--tty',
+            request.param
+        ]).decode().strip()
```
This should be outside the `try`. If this statement fails, `container_id` is not defined and `finally` will fail on a `NameError` exception.

ekohl commented on tests/test_content_client.py:
```diff
@@ -0,0 +1,169 @@
+
+
+@pytest.fixture
+def os_release(katello_client):
```
I'm sure there will come a time when we need to differentiate on major.minor. Let's be explicit that it's only the major version.
```suggestion
def os_release_major(katello_client):
```

ekohl commented on tests/test_content_client.py:
```diff
@@ -0,0 +1,169 @@
+    katello_client.run('yum -y install skopeo')
+    katello_client.run('mkdir containers')
+    katello_client.run_expect([0], f"curl -O http://{registration_hostname}/pub/katello-server-ca.crt")
+    katello_client.run_expect([0], f"skopeo copy {docker_connection} dir:containers --src-creds {user.username}:{user.password} --src-cert-dir .")
+    katello_client.run_expect([1], f"skopeo copy {docker_connection} dir:containers --src-creds doesnotexist:changeme --src-cert-dir .")
+    katello_client.run_expect([1], f"skopeo copy docker://{registration_hostname}/{container_label} dir:containers --src-cert-dir .")
```
I'd swap over these checks: first use invalid credentials, assert that there is nothing in the containers directory, then valid credentials and finally assert that there is content in the containers directory.

evgeni commented:
Needs a rebase, sorry :)

ekohl commented on tests/test_content_client.py:
```diff
@@ -0,0 +1,170 @@
+
+@pytest.fixture
+def katello_agent_client(katello_client, os_release_major):
+    if os_release_major not in ("6", "7"):
+        pytest.skip(f"katello-agent not available on EL{os_release_major}")
```
Is this still true? It's now gotten more complicated.

evgeni replied:
I think this is true enough for most katello releases to be kept as is and we can add the "el8&&k4+" logic at some point later.

ekohl reviewed (approved):
I think this is ready to merge, but let's wait with merging until Katello 4.0.0-rc1 is out.

ehelms commented:
I can hold on merging. Do keep in mind, these client tests will not run unless you configure the set of clients in `variables.json`. So in theory, safe to merge whenever and then do a Forklift update to enable them.

ekohl commented:
> I can hold on merging. Do keep in mind, these client tests will not run unless you configure the set of clients in `variables.json`. So in theory, safe to merge whenever and then do a Forklift update to enable them.

Good point. Let me rephrase it then: let's not enable it in the pipelines until 4.0.0-rc1 is out.

ehelms commented:
Going to merge, and rebase my other PRs and use this as a basis for additional testing and working on updates to pipelines to be able to use this.
